sdk/identity/azure-identity/tests/helpers.py

```diff
@@ -59,53 +58,73 @@ def build_aad_response(  # simulate a response from AAD
 class Request:
     def __init__(
         self,
+        base_url=None,
         url=None,
         authority=None,
         url_substring=None,
         method=None,
-        required_headers={},
-        required_data={},
-        required_params={},
+        required_headers=None,
+        required_data=None,
+        required_params=None,
     ):
         self.authority = authority
+        self.base_url = base_url
         self.method = method
         self.url = url
         self.url_substring = url_substring
-        self.required_headers = required_headers
-        self.required_data = required_data
-        self.required_params = required_params
+        self.required_headers = required_headers or {}
+        self.required_data = required_data or {}
+        self.required_params = required_params or {}
 
     def assert_matches(self, request):
-        # TODO: rewrite this to report all mismatches, and use the parsed url
         url = six.moves.urllib_parse.urlparse(request.url)
 
-        if self.url:
-            assert request.url.split("?")[0] == self.url
-        if self.authority:
-            assert url.netloc == self.authority, "Expected authority '{}', actual was '{}".format(
-                self.authority, url.netloc
-            )
-        if self.url_substring:
-            assert self.url_substring in request.url
-        if self.method:
-            assert request.method == self.method
+        discrepancies = []
+
+        def add_discrepancy(name, expected, actual):
+            discrepancies.append("{}:\n\t expected: {}\n\t   actual: {}".format(name, expected, actual))
+
+        if self.base_url and self.base_url != request.url.split("?")[0]:
+            discrepancies.append("url:\n\texpected base url '{}', actual was '{}'".format(self.base_url, request.url))
+
+        if self.url and self.url != request.url:
+            add_discrepancy("url", self.url, request.url)
+
+        if self.url_substring and self.url_substring not in request.url:
+            discrepancies.append("url: '{}' not in '{}'".format(self.url_substring, request.url))
+
+        if self.authority and url.netloc != self.authority:
```

Comment:
url -> self.url or request.url? (and below)

Reply:
This is correct, if confusing: `url` is `request.url`, parsed; if we were given an expected authority, the actual URL's netloc should match it. `url` is only used here, so I renamed it and moved its declaration closer.